src/pyqasm/visitor.py

```diff
@@ -979,18 +986,91 @@ def _visit_generic_gate_operation(
                     self._function_qreg_transform_map[-1],
                 )
             )
-        # Applying the inverse first and then the power is same as
-        # apply the power first and then inverting the result
+
+        operation.qubits = self._get_op_bits(
+            operation, reg_size_map=self._global_qreg_size_map, qubits=True
+        )
+
+        # ctrl / pow / inv modifiers commute. so group them.
+        exponent = 1
+        ctrl_arg_ind = 0
+        for modifier in operation.modifiers:
+            modifier_name = modifier.modifier
+            if modifier_name == qasm3_ast.GateModifierName.pow and modifier.argument is not None:
+                try:
+                    current_power = Qasm3ExprEvaluator.evaluate_expression(
+                        modifier.argument, reqd_type=qasm3_ast.IntType
+                    )[0]
+                except ValidationError:
+                    raise_qasm3_error(
+                        f"Power modifier argument must be an integer in gate operation {operation}",
+                        span=operation.span,
+                    )
+                exponent *= current_power
+            elif modifier_name == qasm3_ast.GateModifierName.inv:
+                exponent *= -1
+            elif modifier_name in [
+                qasm3_ast.GateModifierName.ctrl,
+                qasm3_ast.GateModifierName.negctrl,
+            ]:
+                try:
+                    count = Qasm3ExprEvaluator.evaluate_expression(
+                        modifier.argument, const_expr=True
+                    )[0]
+                except ValidationError:
+                    raise_qasm3_error(
+                        "Controlled modifier arguments must be compile-time constants "
+                        f"in gate operation {operation}",
+                        span=operation.span,
+                    )
+                if count is None:
+                    count = 1
+                if not isinstance(count, int) or count <= 0:
+                    raise_qasm3_error(
+                        "Controlled modifier argument must be a positive integer "
+                        f"in gate operation {operation}",
+                        span=operation.span,
+                    )
+                # TODO: unroll ctrl qubits
```

Comment:
I think since we are using the `_get_op_bits` in L990, we already have unrolled bits here. You can remove this todo